[Navigation] Introduce navigation-compose-material and support for bottom sheet navigation destinations

jossiwolf opened the pull request:
## Proposed Changes

For easier integration with `compose-material`'s `ModalBottomSheetLayout`, this library provides convenience wrappers and a `BottomSheetNavigator` to be able to declare and navigate to bottom sheet destinations, similar to bottom sheet dialog destinations in vanilla navigation.
`BottomSheetNavigator` drives a `ModalBottomSheetLayout`'s `sheetContent` and state, controlling visibility according to the back stack and composing and disposing destinations that are navigated to or away from. `BottomSheetNavigator` offers the `sheetContent` as a composable that should be set as the `ModalBottomSheetLayout`'s `sheetContent`.

- Composable bottom sheet destinations can be added using `NavGraphBuilder#bottomSheet`
- The BottomSheetNavigator needs to be registered with the `NavController`'s `navigatorProvider`
- Provides a convenience wrapper around `ModalBottomSheetLayout` that accepts a `BottomSheetNavigator`
- Extracted `BackStackEntryIdViewModel` and the `SaveableStateHolder.SaveableStateProvider` extension into their own file in `navigation-compose` because they are needed in `navigation-compose-material`

~~Generating the API files for `navigation-compose-material` makes lint trip up about overloads/Java compatibility on a composable function (`src/main/java/androidx/navigation/compose/material/BottomSheet.kt:35: error: A Kotlin method with default parameter values should be annotated with @JvmOverloads for better Java interoperability`), so any pointers there would be appreciated :)~~ as a workaround I added `navigation-compose-material` to the list of exceptions in `AndroidXExtension#targetsJavaConsumers` locally to generate the API files. That might even be the correct solution, let me know what you think is best.

I also thought about making `navigation-compose-material` a submodule of `navigation-compose` - would that be better?

Please let me know if you have ideas for more tests, I plan on adding a couple more, not super happy with them yet.

Cheers!

## Testing

Test: Added tests for new behavior (`BottomSheetNavigatorTest`, `SheetContentHostTest`, `NavGraphBuilderTest`), ran with `./gradlew test connectedCheck`. More tests to be added.
Demos: Introduced BottomSheetNavDemo

## Issues Fixed

Fixes: b/180247978

jossiwolf committed:
Introduce navigation-compose-material and support for bottom sheet navigation destinations

For easier integration with compose-material's ModalBottomSheetLayout, this library provides convenience wrappers and a BottomSheetNavigator to be able to declare and navigate to bottom sheet destinations, similar to bottom sheet dialog destinations in vanilla navigation.
BottomSheetNavigator drives a ModalBottomSheet's sheetContent and state, controlling visibility according to the back stack and composing and disposing destinations that are navigated to or away from. BottomSheetNavigator offers the sheetContent as a composable that should be set as the ModalBottomSheetLayout's sheetContent.

- Composable bottom sheet destinations can be added using NavGraphBuilder#bottomSheet
- The BottomSheetNavigator needs to be registered with the NavController's navigatorProvider
- Provides a convenience wrapper around ModalBottomSheetLayout that accepts a BottomSheetNavigator
- Extracted BackStackEntryIdViewModel and the SaveableStateHolder.SaveableStateProvider extension into their own file in navigation-compose because they are needed in navigation-compose-material

Test: Added tests for new behavior (BottomSheetNavigatorTest and SheetContentHostTest), ran with ./gradlew test connectedCheck. More tests to be added.
Demos: Introduced BottomSheetNavDemo
Fixes: b/180247978

Change-Id: I51d2c86ce01ef13d68622d368ae1efaee51ffd7c

ianhanniballake commented on buildSrc/src/main/kotlin/androidx/build/LibraryVersions.kt:
```diff
@@ -88,6 +88,7 @@ object LibraryVersions {
     val MEDIA2 = Version("1.2.0-alpha02")
     val MEDIAROUTER = Version("1.3.0-alpha02")
     val NAVIGATION = Version("2.4.0-alpha01")
+    val NAVIGATION_COMPOSE_MATERIAL = Version("1.0.0-alpha01")
```
Navigation is back to using just a single version of all artifacts; we don't want or need a separate version for this artifact (it can just be part of 2.4.0-alpha01 with everything else).

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt:
```diff
@@ -0,0 +1,125 @@
+ * drive the sheet state
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Navigator.Name("ComposeBottomSheetNavigator")
```
Use the same name as the `NavGraphBuilder` extension - `"bottomSheet"`

ianhanniballake commented on navigation/navigation-compose-material/build.gradle:
```diff
@@ -0,0 +1,62 @@
+    implementation(KOTLIN_STDLIB)
+    api(projectOrArtifact(":compose:runtime:runtime"))
+    api(projectOrArtifact(":compose:runtime:runtime-saveable"))
+    api(projectOrArtifact(":navigation:navigation-common-ktx"))
```
You don't need this dependency as it is transitively included in `navigation-compose`.

ianhanniballake commented on navigation/navigation-compose-material/build.gradle:
```diff
@@ -0,0 +1,62 @@
+}
+
+androidx {
+    name = "Compose Material Navigation"
```
The name should be changed to Navigation Compose Material (we always refer to it as Navigation Compose)

ianhanniballake commented on navigation/navigation-compose-material/build.gradle:
```diff
@@ -0,0 +1,62 @@
+    mavenVersion = LibraryVersions.NAVIGATION_COMPOSE_MATERIAL
+    mavenGroup = LibraryGroups.NAVIGATION
+    inceptionYear = "2021"
+    description = "Compose Material integration with Navigation"
```
Reverse the ordering here as well - Navigation integration with Compose Material

ianhanniballake commented on navigation/navigation-compose-material/build.gradle:
```diff
@@ -0,0 +1,62 @@
+    inceptionYear = "2021"
+    description = "Compose Material integration with Navigation"
+    runApiTasks = new RunApiTasks.Yes()
+    legacyDisableKotlinStrictApiMode = true
```
Generally for Navigation artifacts, we've enabled strict API mode (it has caught a number of bugs for us). I'm personally not particularly concerned about visibility issues in our androidTest code so feel free to either ignore or just mark them all as public as needed (it doesn't actually matter).

jossiwolf replied:
Sounds good! Heh maybe that was causing the issues with generating the lint baseline? 🤔 (yes, looks like it actually was!)

Looks like it's [disabled in `navigation-compose`](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:navigation/navigation-compose/build.gradle;l=62) too, maybe we should remove that there too?

I saw that the Compose libraries apply the Kotlin plugin through `AndroidXUiPlugin.applyAndConfigureKotlinPlugin(project)`, I think it'd be good to use that too! There were a couple of lint checks (composable naming) it was complaining about so that'll be nice. Let me know if we shouldn't do this!

Looking at the Compose libraries, they have set `legacyDisableKotlinStrictApiMode = true` too 🤔 I'll just ignore the visibility warnings in tests for now

ianhanniballake commented on navigation/navigation-compose-material/integration-tests/navigation-demos/build.gradle:
```diff
@@ -0,0 +1,47 @@
+}
+
+androidx {
+    name = "Compose Material Navigation Demos"
```
Ditto on reordering the names to Navigation Compose Material Demos

ianhanniballake commented on navigation/navigation-compose-material/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/material/demos/BottomSheetNavDemo.kt:
```diff
@@ -0,0 +1,96 @@
+fun BottomSheetNavDemo() {
+    val navController = rememberNavController()
+    val bottomSheetNavigator = rememberBottomSheetNavigator()
+    navController.navigatorProvider += bottomSheetNavigator
```
This is going to cause issues if this method recomposes. Perhaps what we actually want here is a `rememberNavController()` overload (/optional parameter) that takes a `NavHostController.() -> Unit` lambda where we could do this type of configuration when the NavController is created?

I'm not sure how we'd effectively match the `remember` scoping of the `NavController` and the `BottomSheetNavigator`, as generally we'd only want those to be 1:1. Perhaps not really an issue since the `remember` is ~unconditional.

jossiwolf replied:
Yeah, good catch! I think an optional parameter would be nice.

```kotlin
val bottomSheetNavigator = rememberBottomSheetNavigator()
val navController = rememberNavController {
    navigatorProvider += bottomSheetNavigator
}
```

```kotlin
@Composable
fun rememberNavController(
    configureNavController: NavHostController.() -> Unit = { }
): NavHostController {
    val context = LocalContext.current
    return rememberSaveable(configureNavController, saver = NavControllerSaver(context)) {
        createNavController(context).apply(configureNavController)
    }
}
```

I don't think that scoping is a big issue here, and making `configureNavController` `@Composable` isn't great either imo. In case it's an issue, we'll see it pop up in alpha I think :)

If you agree on that approach, I'll be happy to put this up in another PR!

Edit: Mhm, I'm not happy with that - I guess the configuration lambda would have to be composable after all? Let me know what your thoughts are. Just questioning how we'd track changes to anything that's referenced in the lambda.

ianhanniballake commented on navigation/navigation-compose-material/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/material/demos/BottomSheetNavDemo.kt:
```diff
@@ -0,0 +1,96 @@
+    Button(onClick = showAnotherSheet) {
+        Text("Click me to show another sheet!")
+    }
+}
```
Make sure to add a newline at the bottom of every new file.

ianhanniballake commented on navigation/navigation-compose-material/samples/build.gradle:
```diff
@@ -0,0 +1,48 @@
+}
+
+androidx {
+    name = "AndroidX Compose UI Navigation Material Integration Samples"
```
Ditto on the name and description.

ianhanniballake commented on navigation/navigation-compose-material/src/androidTest/java/androidx/navigation/compose/material/SheetContentHostTest.kt:
```diff
@@ -0,0 +1,131 @@
+            .containsExactly(backStackEntry)
+    }
+
+
```
Nit: Make sure to clean up multiple empty lines in a row.

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt:
```diff
@@ -0,0 +1,125 @@
+ * @param sheetState The [ModalBottomSheetState] that the [BottomSheetNavigator] will use to
+ * drive the sheet state
+ */
+@OptIn(ExperimentalMaterialApi::class)
```
We'll need to follow up the Compose team - while we can use experimental APIs in alpha, policy states that we **cannot** move to beta with APIs that depend on experimental APIs from another artifact group. We'll need to get agreement that these APIs will become stable before Compose 1.0 before we can add APIs built on top of them.

I can do most of this work if you'd like - I'd appreciate a list of APIs we'll need to implement this fully (or if there's some minimal set we can use and leave helpers that need more experimental APIs as demos / docs only).

jossiwolf replied:
Oh, interesting. That makes sense and is probably for the better 😄 

I think there are quite a few issues around bottom sheets and the swipeable APIs and I'm not sure what the plan is, but my understanding was that these might well stay experimental until after 1.0. Looking at the APIs we're using though - I think it might be possible to make them stable. At least `ModalBottomSheetState` can probably be made stable by moving the `SwipeableState` into the class and not inheriting from it anymore (same as with `DrawerState`, https://android-review.googlesource.com/c/platform/frameworks/support/+/1613082/).

If you could talk to the team about this, that'd be awesome! (You can also make a group chat with Matvei, you and me if you want)

The APIs we need are
- `ModalBottomSheetLayout` (though we only need that for our convenience method that's `@ExperimentalMaterialApi` anyways)
- `ModalBottomSheetState` for use in `BottomSheetNavigator` and `SheetContentHost`

The APIs we need from `ModalBottomSheetState` are all related to visibility but that's pretty much at the core of the swipeable issues anyways.

ianhanniballake reviewed (changes requested):
Just a high level review at the moment - I hope to have some time to dive into the details in a later review (one CL huh? lol).

It looks like there's a few more infrastructural things (like the `SaveableStateProvider`) that we should split off into separate PRs that can be merged in prior to this. Let me know if you'd like to do this or if you'd like us to do this ahead of time.

I have some ideas on how we'd like to approach each of these if you'd like to reach out to me or just list out what dependencies you need here and I can comment on each.

jossiwolf committed:
Document BottomSheetNavigator's behavior with nested flows more closely

Change-Id: I4797acc9652240fc0804cc7ff1e66a4c012206c2

jossiwolf committed:
Clean up tests

Change-Id: Ib8df7d5251eb998d8fd4b525c211eb21910c62c2

jossiwolf committed:
Enable strict API mode

Change-Id: Ia18698d028a7001205a98be7248a04b55abe5826

jossiwolf committed:
Use NAVIGATION's version for navigation-compose-material

Change-Id: Ia219067c52498f385057488a97f1a956e77cc620

jossiwolf committed:
Rename "Compose Material Navigation" -> "Navigation Compose Material"

Change-Id: I6f9fb474aa7da3f157a06273247ecb0b4e97cfb1

jossiwolf committed:
Update BottomSheetNavigator's Navigator.Name to "BottomSheetNavigator" to match our NavGraphBuilder extension

Change-Id: I7b78ac5a908391a9283014aaa5b303b49d454810

jossiwolf committed:
Remove unneeded dependency

Change-Id: Ib3321c5aac19d364a023ef11f65164b2bd930f41

jossiwolf committed:
Lint & Formatting

Change-Id: I18fb46962a7fdfd1f491e89502c723e341b0291f

jossiwolf committed:
Apply Kotlin plugin through AndroidXUiPlugin#applyAndConfigureKotlinPlugin instead of applying it directly

This way, we benefit from lint suppressions for compose-specific warnings

Change-Id: I7db78451e86a89143ebfe64cea08f541a7535a93

jossiwolf committed:
Add testOnSheetDismissedCalled_programmaticDismiss to test programmatic sheet dismissal

Change-Id: I5d115c9f1e59530cec3e0e630d71776a721c2e50

jossiwolf committed:
Remove ActiveNavBackStackEntry.kt

Change-Id: I5d2767396f1307c9dd6d29d5005bec3a24d371ad

jossiwolf committed:
Manually apply Kotlin Plugin for now

I was having issues in playground mode where the COMPOSE_MPP_ENABLED flag was inferred as true

Change-Id: I3c410387c5268a6bd4cc22c99aa51b9b3fb87c5b

jossiwolf committed:
Use runBlockingTest for SheetContentHostTest test methods

Running into quite a few weird things w/ Coroutines and dispatchers...

Change-Id: If3716d5f7b8d3db1b95b00176ecff985022feef3

jossiwolf committed:
Use new LocalOwnersProvider API and lift SaveableStateHolder out of individual NavBackStackEntry-scoped composable to hoist it

Change-Id: I9d600197acc502a1326cf2520801e7ee3f04b218

jossiwolf committed:
Use public APIs for routes

Change-Id: Ie3ff2205d262a714827c4422df92457dbbc5c9c0

jossiwolf committed:
Remove unneeded import

Change-Id: I691bd71529919b2397139522147402bcc62bd25c

jossiwolf committed:
Add NavGraphBuilderTest.kt to test the NavGraphBuilder#bottomSheet extension

These tests are pretty much the same as navigation-composes' NavGraphBuilderTest

Change-Id: I7924e2c9722d250a83ff96019d3d5cce60fe2f7a

jossiwolf committed:
Register navigator with navController as side effect

Change-Id: I11613c037ce2ade06a766139e23b9ded36137f34

jossiwolf committed:
Remove old BackStackEntryIdViewModel API entry; we already have the change upstream

Change-Id: I0d246f19e302a6382b63013df842ec7d5d42ddc8

jossiwolf committed:
Add test case for dismiss callback when ModalBottomSheetLayout starts out in the `EXPANDED` state

Change-Id: I2ccea6ce2fd80681a1c0f731a159387d26271444

jossiwolf committed:
Add temporary workaround to pop back stack on sheet dismiss

NavigatorState's pop currently can't be called outside of popBackStackEntry; we need to pop the back stack outside of that though, e.g. when the sheet is no longer visible. BottomSheetNavigator now takes a onSheetDismissed lambda param that can encapsulate that logic so that the navigator doesn't have to depend on the NavController.

For convenience, rememberBottomSheetNavigator takes a NavController and calls NavController#popBackStack in the onSheetDismissed callback.

This will be removed ASAP; probably Navigation 2.4.0-alpha02. We'll provide a ReplaceWith quickfix then too.

See b/187873799 for a few more details on the issue.

Change-Id: Ia5da50279ef66dedea5b111638f336771574859a

ianhanniballake commented on navigation/navigation-compose-material/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/material/demos/BottomSheetNavDemo.kt:
```diff
@@ -0,0 +1,97 @@
+            }
+            composable(Destinations.Feed) { Text("Feed!") }
+            bottomSheet(Destinations.Sheet + "?arg={arg}") { backstackEntry ->
+                val arg = backstackEntry.arguments?.get("arg") as String
```
Nit: use `getString("arg")` rather than casting.

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt:
```diff
@@ -0,0 +1,149 @@
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+public fun rememberBottomSheetNavigator(
+    sheetState: ModalBottomSheetState =
+        rememberModalBottomSheetState(ModalBottomSheetValue.Hidden),
```
These lines brought up an interesting question I had:
- We need to allow hoisting this state to allow for coordinating other things with the bottom sheet, is that correct? Or is there some other reason?
- What happens when someone calls `hide()` or `show()` on this state externally from the `BottomSheetNavigator`? What does that do to the `BottomSheetNavigator`?
- What happens after a configuration change? With an initial state of always `ModalBottomSheetValue.Hidden`, does that mean that when the `NavController` restores its state that the bottom sheet animates in again? I guess we don't know the 'intial state' until the `BottomSheetNavigator` has its `onAttach()` called and we can check the `NavigatorState.backStack` to see if it empty or not? I don't see any 'snap to state' kind of API?

jossiwolf replied:
Great questions!

1. Yes! There might be cases where you'd want to hide or show the sheet (though I imagine most of these use cases would be covered by simply navigating) manually. Theoretically, we do not need to offer this control (it would save us some headache when thinking about the initial bottom sheet state; what if somebody set it to `Expanded` here?) though. I'd be open to removing this param! It would help prevent misuse and the `sheetState` can be accessed through the navigator in any case. I'll play around with that a little bit :)
2. The only way we can react to `show`/`hide` calls is by looking at the `sheetVisiblity` - we currently have a `snapshotFlow` observing the visibility and calling a callback (that will then pop the back stack) when the sheet was visible and changed its state to hidden. So that `LaunchedEffect`/`snapshotFlow` covers both manual and programmatic dismiss (we have tests for that in `SheetContentHostTest` too!).  `show()`ing the sheet is a good question... I thought about it but didn't find any good solution. Calling `show()` is a little ambiguous: 
    1. It implies the sheet is currently hidden and the current nav destination is not a `BottomSheetNavigator.Destination`
    2. When you call `show()`, what is the expected outcome? What should we show?
Do you think we should just ignore `show` calls? Or any other ideas?
3. `rememberModalBottomSheetState` uses `rememberSaveable` under the hood so we will stay in the same state across config changes. But we might not actually want that as it means that the sheet and navigator might be out of sync. I think it'd be better to react to that after the config change though, when we actually know if we want to show the sheet or not.
> snap to state API

Is that in the navigator?

jossiwolf replied:
(also, if you're asking these questions, it makes sense to document that - I'll add explanations around this to the docs later on!)

jossiwolf replied:
Sooo
1. I just went back and forth on this - IMO we should keep exposing this because folks might want to observe the state to run things etc. Maybe a point can be made that this should be done using effects so I'm not entirely against removing it and adding it back if requested.
2. I'm still not sure what we should do when somebody calls `show` - we could just ignore it. Feels like a sensible thing to do to me. Wouldn't be an issue if we weren't hoisting the state.
3. Is working now 🎉 Working on a test

jossiwolf committed:
Experiment with custom-restoring state

Change-Id: I5575a20225ada25bb46bc4e7b4f28002aafb299b

chrisfillmore commented on navigation/navigation-compose-material/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/material/demos/BottomSheetNavDemo.kt:
```diff
@@ -0,0 +1,97 @@
+fun BottomSheetNavDemo() {
+    val navController = rememberNavController()
+    val bottomSheetNavigator = rememberBottomSheetNavigator(navController = navController)
+    SideEffect { navController.navigatorProvider += bottomSheetNavigator }
```
Hello, I hope you don't mind a question from a stranger, but I am working on a similar problem and was referred to your PR as an example.

My question is, is the use of `SideEffect` correct here? Since the block doesn't run until after composition, if you call `NavGraphBuilder.bottomSheet()` immediately, you will get an IllegalStateException. That is to say, I believe the following hypothetical code would fail:

```Kotlin
val navController = rememberNavController()
val bottomSheetNavigator = rememberBottomSheetNavigator(navController = navController)
SideEffect { navController.navigatorProvider += bottomSheetNavigator }
// Note: not wrapped in ModalBottomSheetLayout
NavHost(...) {
  bottomSheet(...) {
    ...
  }
}
```
It would fail with this error:
> java.lang.IllegalStateException: Could not find Navigator with name "BottomSheetNavigator". You must call NavController.addNavigator() for each navigation type.

If `SideEffect` were removed, there would be no error. However I am not sure if it is correct to remove `SideEffect`. Perhaps @ianhanniballake can advise?

I speculate that you haven't seen this error so far, because the ModalBottomSheetLayout begins in "Hidden" state and is thus not part of the composition, thus the `SideEffect` is run before `bottomSheet()`. I suspect the tests pass because they don't use `SideEffect`. But I am speculating; I haven't run the code.

ianhanniballake replied:
`SideEffect` is specifically exactly what should be used, as explained in [b/187443146](https://issuetracker.google.com/187443146) and specifically made possible in #176. The navigation graph is applied as part of a `DisposableEffect`, so this all works perfectly in practice 😄

ianhanniballake replied:
Given the discussion in [b/193555369](https://issuetracker.google.com/issues/193555369), I think I was wrong here and that `SideEffect` is not the right choice. As per the [third comment](https://issuetracker.google.com/issues/193555369#comment3), it is safe to apply the Navigator as part of composition at this point, so let's do that until the better API mentioned in that bug is created.

jossiwolf committed:
Use AndroidXComposePlugin instead of AndroidXUiPlugin

Change-Id: I1315a295b8c809dd2d6eb5ef5d79275648881bb9

jossiwolf committed:
Switch to NavigatorState#pop over popBackStack

Change-Id: I415f0a2877654513126ba13d3b5d60e90ebf5675

jossiwolf committed:
Revert "Experiment with custom-restoring state"

Keeping d67415c6 in the commit history but looking at other approaches now!

This reverts commit d67415c6

Change-Id: I35e3dc49417fed7380b723c0cd99bf3d88533e79

jossiwolf committed:
Access NavigatorState's back stack directly

https: //android-review.googlesource.com/c/platform/frameworks/support/+/1768966 (08ddc8b8) enables doing that because now the graph gets set in the composition, meaning we already have access to it in the first composition and not only in the second.
Change-Id: I11e0184aa9365ca72ee82d7af2ec722b12ca2ffb

jossiwolf committed:
Use transition APIs

Change-Id: Iacad03dc37a8e29d86dc582c76d282b71ee60ad4

jossiwolf committed:
Move sheet dismissal listener logic to SheetContentHost.kt

Change-Id: Id9569c55d598a574e2675b0018b3bf5013d649cb

jossiwolf committed:
Remove BottomSheetNavigator's onBottomSheet param

Change-Id: I88de02fcc0d6fa1decda0a72c7308e37320d5542

jossiwolf committed:
Revert "Move sheet dismissal listener logic to SheetContentHost.kt"

We won't use popWithTransition because tracking the sheet's state changes isn't awesome and we only really want to pop once the transition is finished since transitions aren't cancellable currently.

This reverts commit d55fb2e1

Change-Id: I9d7946677a9a5b31b3a1c124e32aab9674fd2274

jossiwolf committed:
Retrieve argument with getString instead of casting

Change-Id: I8de67c7203f583e55255c1d2b5c430d1d9b32b50

jossiwolf committed:
Make SheetContentHost's onSheetShown param non-optional

Change-Id: Icdcc6fb9d8a2dc515988d4b2a96b064417868b49

jossiwolf committed:
Add onSheetShown test

Change-Id: I809f44d11207dd58cd2a3065cfb1af101c5637d4

jossiwolf committed:
Add state restoration test

It's currently failing and I can't quite figure out why yet :/

Change-Id: I9b55fda97576b92916982fc409268aea78f1b0f0

jossiwolf committed:
`@Ignore` testSheetShownAfterNavControllerRestoresState

Change-Id: I0f7e9f1d4d72c824977052d465cf45004a26cec9

jossiwolf committed:
Use rememberNavController over manually creating NavHostController because of crash

Change-Id: I96cc1cf2eed55c5194905c41d2278de40fcaed4b

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt:
```diff
@@ -0,0 +1,129 @@
+        navigatorExtras: Extras?
+    ) {
+        entries.forEach { entry ->
+            state.pushWithTransition(entry)
```
If we're `pushWithTransition` multiple entries (either because multiple entries are passed to this method or `navigate` is called multiple times in a row (which is probably the more likely case - that's how deep linking works), but we only display the latest one, how do those other entries get `onTransitionComplete` called?

jossiwolf replied:
Done in e2d181a

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt:
```diff
@@ -0,0 +1,129 @@
+        // We always replace the sheet's content instead of overlaying and nesting floating
+        // window destinations. That means that only *one* concurrent destination is supported by
+        // this navigator.
+        val latestEntry = backStackEntries.lastOrNull { entry ->
```
So when the system back button is pressed and `NavController` calls `popBackStack()` on the Navigator, how does this Navigator handle that? The default behavior of `popBackStack()` will call `pop`, which will immediately destroy the state of the `NavBackStackEntry` and remove it from `backStackEntries`. Should you be calling `popWithTransition()` there and looking at `transitionsInProgress`.lastOrNull()` and `backStackEntries` here to know what to be hiding?

jossiwolf replied:
So when the entry gets removed from `backStackEntries`, the `SheetContentHost` will be recomposed with the new entry or `null`, indicating that the sheet should be empty. `SheetContentHost` has a `DisposableEffect` keyed to its `backStackEntry` to hide the sheet whenever this sheet content leaves the composition - so effectively, the sheet is hidden after the entry has been destroyed. Writing that, yeah sounds like we should use `popWithTransition` for `popBackStack`.

I thought about

```kotlin
override fun popBackStack(popUpTo: NavBackStackEntry, savedState: Boolean) {
    state.popWithTransition(popUpTo, savedState)
}
```

```kotlin
SheetContentHost(
   ...,
   onSheetDismissed = { backStackEntry ->
      transitionsInProgress[backStackEntry]?.onTransitionComplete() ?: state.pop(backStackEntry, saveState = false)
   }
)
```

That'd solve the issue of always popping with a transition and give us the possibility to fall back on `pop`, i.e. when the sheet is dismissed by a user gesture.

Is that what you had in mind too?

ianhanniballake replied:
That seems reasonable to me, yeah.

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/SheetContentHost.kt:
```diff
@@ -0,0 +1,130 @@
+
+        // Whenever the composable associated with the latestEntry enters the composition, we
+        // want to show the sheet, and hide it when this composable leaves the composition
+        DisposableEffect(Unit) {
```
How does this handle navigating from one bottom sheet destination to another? I would have thought you'd need a `DisposableEffect(backStackEntry` or something similar here? When navigating between bottom sheets, what is the expected behavior here? I would think the first one would want to hide completely, then the second one shows? I'm not sure how to write that logic though.

jossiwolf replied:
Yeah you're right about that, the effect should be scoped to the entry. The sheet would hide and then a new one appear - I'm not suuuuper happy with that because if you think about nesting destinations in a sheet you'll most likely want to have a transition along the x-axis. But I think that's out of scope for this PR :D

ianhanniballake replied:
Well, navigating to a new bottom sheet destination is not about replacing the content within the bottom sheet, but hiding the existing bottom sheet destination and then showing a brand new bottom sheet. That's the styling we should shoot for.

jossiwolf replied:
Yup. Would be cool to enable something else in the future but we should stick to "hide then show" for now (that's also what it is)

ianhanniballake commented on navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt:
```diff
@@ -0,0 +1,129 @@
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+public fun rememberBottomSheetNavigator(
+    sheetState: ModalBottomSheetState =
```
I'm still not entirely on board with having this an an input given that we specifically don't want them to set a different `initialValue` or `confirmStateChange`, not to mention wanting to avoid the calls to `show()` and `hide()` externally.

Perhaps we can expose the public properties of `ModalBottomSheetState` as part of `BottomSheetNavigator`'s API surface, thus allowing developers to use the Navigator directly where they were using the state while maintaining the ownership properties we need.

jossiwolf replied:
Yeah I agree. Done in d0a03a0, would love to get your thoughts about using an extra class to represent that. Adding these to the navigator felt a bit bloaty, but discoverability is worse this way.

jossiwolf committed:
Use popWithTransition

Change-Id: I149c313a44df7917e648548f01b2a09bf498281d

jossiwolf committed:
Make sheetState an impl detail of BottomSheetNavigator.kt and expose select properties through BottomSheetNavigatorState

Change-Id: Icdf496e0b68256034b25781f9d3254142807a686

jossiwolf committed:
Use markTransitionComplete API

Change-Id: I5ab782fa46d0b02b55c82e69e5098c0ac25428be

jossiwolf committed:
Pin Compose 1.0.1 for navigation-compose-material

Change-Id: I945813b06cb945533fc0699bb5b54732da16e5e0

jossiwolf committed:
Use internalShow instead of show, the show call is cancelled

Change-Id: Ic7f6020509378d729d9e906235c2cba5d6c7da0f

jossiwolf committed:
Pop only dismissed sheet's entry

Change-Id: I6017c5bfbbd438f92fc9ad36a3c09de1f5768ad5

jossiwolf committed:
Mark all entries' transitions as complete

Change-Id: I1eb083a3c5b81485bdb1c33d2c8c6a07dbce82b0

jossiwolf committed:
Attach NavigatorState before first composition

In this test, we were composing (and thus trying to access the BottomSheetNavigator's state) before we had attached the state to the navigator

Change-Id: I63309dd8b299ffafaa571226b091eb62ba1f3584

jossiwolf committed:
Lint :)

Change-Id: I9bb5d36316c892930d742a72fcfd0e8b9b191dd9

jossiwolf committed:
Add `testNavigateCompletesEntriesTransitions` Test

Change-Id: If594ba4d1d5c884bbf8ef9f723e4a00c28e0ae62

jossiwolf committed:
Make BottomSheetNavigator's fields public

Change-Id: I838fd686b37624f6f2e9da9b7da15fdec9b3f6ca

ianhanniballake commented on navigation/navigation-compose-material/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/material/demos/BottomSheetNavDemo.kt:
```diff
@@ -0,0 +1,97 @@
+fun BottomSheetNavDemo() {
+    val navController = rememberNavController()
+    val bottomSheetNavigator = rememberBottomSheetNavigator()
+    SideEffect { navController.navigatorProvider += bottomSheetNavigator }
```
Just a reminder to remove the `SideEffect` here.

ianhanniballake commented on navigation/navigation-compose-material/src/androidTest/java/androidx/navigation/compose/material/NavGraphBuilderTest.kt:
```diff
@@ -0,0 +1,171 @@
+        val arg = "myarg"
+        composeTestRule.setContent {
+            navController = TestNavHostController(LocalContext.current)
+            navController.navigatorProvider.addNavigator(createBottomSheetNavigator())
```
You can use `+=` instead of `addNavigator` here and elsewhere

ianhanniballake commented on navigation/navigation-compose-material/src/androidTest/java/androidx/navigation/compose/material/NavGraphBuilderTest.kt:
```diff
@@ -0,0 +1,171 @@
+
+        composeTestRule.runOnUiThread {
+            assertWithMessage("Destination should be added to the graph")
+                .that(firstRoute in navController.graph)
```
`firstRoute` is the route of the navigation graph and not the `bottomSheet` destination itself - don't we want to check on the existence of `secondRoute`? That seems to imply that this check should be something like `secondRoute in (navController.graph.findNode(firstRoute) as NavGraph)`?

jossiwolf replied:
Yep good catch!

ianhanniballake commented on navigation/navigation-compose-material/src/androidTest/java/androidx/navigation/compose/material/NavGraphBuilderTest.kt:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Test
+    fun testNavigationNestedInGraph() {
```
TBH, I'm not sure if this adds anything over the previous test, maybe they can be combined? The difference between start destination and not start destination isn't related to the `bottomSheet` extension method.

jossiwolf replied:
Ya I agree! I don't think it's needed, removed it :)

ianhanniballake reviewed (commented):
I'm really happy with how this works.

While we can't approve this due to the usage of experimental APIs across different artifact groups, I'd be absolutely happy with this as part of Accompanist or as a standalone library that anyone using Navigation Compose can use.

jossiwolf committed:
Add BottomSheetNavigator to NavController in composition

Change-Id: I1f03e990370dc421cca089f65d24399446979eb3

jossiwolf committed:
Remove `testNavigationNestedInGraph` test

Change-Id: I4ae5d166118c55ec34d9a635660521d10b9afdc8

jossiwolf commented:
Ayyyy whohoo! Thank you so much for sticking with this PR :) Gonna move over to Accompanist!

(btw, checks are failing because of Metalava. Hope it'll just magically work over there :D)
